File: pintia/gplt_practice/l2/_007.java
import java.io.*;
import java.util.*;

public class _007 {
    static class Info implements Comparable<Info> {
        int cnt = 0, area = 0, size = 0, mn = 0;
        double avg_cnt, avg_area;

        @Override
        public int compareTo(Info o) {
            return avg_area == o.avg_area ? Integer.compare(mn, o.mn) : Double.compare(o.avg_area, avg_area);
        }
    }

    static final int N = 1010, M = 10010;
    static int n;
    static int[] roots = new int[M];
    static int[] sizes = new int[M];
    static Set<Integer> ids = new HashSet<>();
    static Map<Integer, Info> infos = new HashMap<>();

    static int find(int x) {
        return x == roots[x] ? x : (roots[x] = find(roots[x]));
    }

    static void join(int p, int q) {
        roots[find(p)] = find(q);
    }

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken(); n = (int) in.nval;
        for (int i = 1; i < M; ++i) roots[i] = i;
        for (int i = 0; i < n; ++i) {
            in.nextToken(); int id = (int) in.nval;
            in.nextToken(); int fa = (int) in.nval;
            in.nextToken(); int mo = (int) in.nval;
            in.nextToken(); int k = (int) in.nval;
            ids.add(id);
            if (!infos.containsKey(id)) infos.put(id, new Info());
            infos.get(id).mn = id;
            if (fa != -1) {
                join(id, fa);
                ids.add(fa);
                if (!infos.containsKey(fa)) infos.put(fa, new Info());
                infos.get(fa).mn = fa;
            }
            if (mo != -1) {
                join(id, mo);
                ids.add(mo);
                if (!infos.containsKey(mo)) infos.put(mo, new Info());
                infos.get(mo).mn = mo;
            }
            while (k-- > 0) {
                in.nextToken();
                int kid = (int) in.nval; ids.add(kid);
                if (!infos.containsKey(kid)) infos.put(kid, new Info());
                infos.get(kid).mn = kid;
                join(id, kid);
            }
            in.nextToken(); infos.get(id).cnt = (int) in.nval;
            in.nextToken(); infos.get(id).area = (int) in.nval;
        }
        List<Integer> lst = new ArrayList<>();
        for (int id : ids) {
            int root = find(id);
            if (root == id) {
                lst.add(id);
            } else {
                infos.get(root).cnt += infos.get(id).cnt;
                infos.get(root).area += infos.get(id).area;
            }
            infos.get(root).mn = Math.min(infos.get(root).mn, id);
            ++infos.get(root).size;
        }
        List<Info> ans = new ArrayList<>();
        for (int id : lst) {
            Info curr = infos.get(id);
            curr.avg_area = curr.area * 1.0 / curr.size;
            curr.avg_cnt = curr.cnt * 1.0 / curr.size;
            ans.add(curr);
        }
        StringBuilder sb = new StringBuilder();
        sb.append(ans.size()).append('\n');
        Collections.sort(ans);
        for (Info u : ans) {
            sb.append(String.format("%04d %d %.3f %.3f\n", u.mn, u.size, u.avg_cnt, u.avg_area));
        }
        System.out.print(sb);
    }
}

File: pintia/gplt_practice/l2/_007.cpp
#include <cstdio>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

const int N = 1010, M = 10010;
int n;
int roots[M], sizes[M];
unordered_set<int> ids;

struct Info {
    int cnt, area, size, mn;
    double avg_cnt, avg_area;

    bool operator<(const Info& o) const {
        return avg_area == o.avg_area ? mn < o.mn : avg_area > o.avg_area;
    }
};
unordered_map<int, Info> infos;

int find(int x) {
    return x == roots[x] ? x : (roots[x] = find(roots[x]));
}

void join(int p, int q) {
    roots[find(p)] = find(q);
}

int main() {
    scanf("%d", &n);
    for (int i = 1; i < M; ++i) roots[i] = i;
    for (int i = 0; i < n; ++i) {
        int id, fa, mo, k, t;
        scanf("%d%d%d%d", &id, &fa, &mo, &k);
        ids.insert(id);
        infos[id].mn = id;
        if (~fa) join(id, fa), ids.insert(fa), infos[fa].mn = fa;
        if (~mo) join(id, mo), ids.insert(mo), infos[mo].mn = mo;
        while (k--) {
            int kid; scanf("%d", &kid);
            ids.insert(kid);
            infos[kid].mn = kid;
            join(id, kid);
        }
        scanf("%d%d", &infos[id].cnt, &infos[id].area);
    }
    vector<int> vec;
    for (auto& id : ids) {
        int root = find(id);
        if (root == id) vec.push_back(id);
        else {
            infos[root].cnt += infos[id].cnt;
            infos[root].area += infos[id].area;
        }
        infos[root].mn = min(infos[root].mn, id);
        ++infos[root].size;
    }
    vector<Info> ans;
    for (auto& id : vec) {
        Info& curr = infos[id];
        curr.avg_area = curr.area * 1.0 / curr.size;
        curr.avg_cnt = curr.cnt * 1.0 / curr.size;
        ans.push_back(curr);
    }
    printf("%d\n", ans.size());
    sort(ans.begin(), ans.end());
    for (auto& u : ans) {
        printf("%04d %d %.3lf %.3lf\n", u.mn, u.size, u.avg_cnt, u.avg_area);
    }

    return 0;
}